Compiler crashes when a private early initializer is referenced in an inner class

Reproducer:

```
class Parent {
    private val field: Int = 3
}

class Child(n: Int) extends {
    private val field = n
} with Parent {
    class Inner {
        def f = field
    }
}
```

Here's the stack trace:

```
error: java.lang.AssertionError: assertion failed:
  expected exactly one field by name field in List(private[this] val field: Int = _) of class Child's early initializers
     while compiling: Test.scala
        during phase: constructors
     library version: version 2.12.3
    compiler version: version 2.12.3
  reconstructed args: -Ystop-before:jvm

  last tree to typer: This(class Inner)
       tree position: line 11 of Test.scala
            tree tpe: Child#Inner
              symbol: class Inner in class Child
   symbol definition: class Inner extends Object (a ClassSymbol)
      symbol package: <empty>
       symbol owners: class Inner -> class Child
           call site: class Child in package <empty>

== Source file context for tree position ==

     8     private val field = n
     9 } with Parent {
    10     class Inner {
    11         def f = field
    12     }
    13 }
    14
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.$anonfun$triage$1(Constructors.scala:613)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.triage(Constructors.scala:607)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.<init>(Constructors.scala:591)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer.transformed(Constructors.scala:694)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.$anonfun$transform$1(Constructors.scala:85)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:85)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:29)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2589)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
        at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1437)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
        at scala.reflect.internal.Trees.itransform(Trees.scala:1437)
        at scala.reflect.internal.Trees.itransform$(Trees.scala:1347)
        at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
        at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
        at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:42)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:89)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:29)
        at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
        at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:426)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:419)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:390)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:390)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:390)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1431)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1416)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1412)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1515)
        at scala.tools.nsc.Driver.doCompile(Driver.scala:35)
        at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
        at scala.tools.nsc.Driver.process(Driver.scala:55)
        at scala.tools.nsc.Driver.main(Driver.scala:68)
        at scala.tools.nsc.Main.main(Main.scala)
Exception in thread "main" java.lang.AssertionError: assertion failed:
  expected exactly one field by name field in List(private[this] val field: Int = _) of class Child's early initializers
     while compiling: Test.scala
        during phase: constructors
     library version: version 2.12.3
    compiler version: version 2.12.3
  reconstructed args: -Ystop-before:jvm

  last tree to typer: This(class Inner)
       tree position: line 11 of Test.scala
            tree tpe: Child#Inner
              symbol: class Inner in class Child
   symbol definition: class Inner extends Object (a ClassSymbol)
      symbol package: <empty>
       symbol owners: class Inner -> class Child
           call site: class Child in package <empty>

== Source file context for tree position ==

     8     private val field = n
     9 } with Parent {
    10     class Inner {
    11         def f = field
    12     }
    13 }
    14
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.$anonfun$triage$1(Constructors.scala:613)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.triage(Constructors.scala:607)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer$Triage.<init>(Constructors.scala:591)
        at scala.tools.nsc.transform.Constructors$TemplateTransformer.transformed(Constructors.scala:694)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.$anonfun$transform$1(Constructors.scala:85)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:85)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:29)
        at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2589)
        at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
        at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1437)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
        at scala.reflect.internal.Trees.itransform(Trees.scala:1437)
        at scala.reflect.internal.Trees.itransform$(Trees.scala:1347)
        at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
        at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
        at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:42)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
        at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:89)
        at scala.tools.nsc.transform.Constructors$ConstructorTransformer.transform(Constructors.scala:29)
        at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
        at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:426)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:419)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:390)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:390)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:390)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1431)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1416)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1412)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1515)
        at scala.tools.nsc.Driver.doCompile(Driver.scala:35)
        at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
        at scala.tools.nsc.Driver.process(Driver.scala:55)
        at scala.tools.nsc.Driver.main(Driver.scala:68)
        at scala.tools.nsc.Main.main(Main.scala)
```

--- a/src/compiler/scala/tools/nsc/transform/Constructors.scala
+++ b/src/compiler/scala/tools/nsc/transform/Constructors.scala
@@ -14,7 +14,7 @@ package scala.tools.nsc
 package transform
 
 import scala.annotation._
-import scala.collection.mutable
+import scala.collection.mutable, mutable.ListBuffer
 import scala.reflect.internal.util.ListOfNil
 import scala.tools.nsc.Reporting.WarningCategory
 import symtab.Flags._
@@ -29,8 +29,7 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
   /** the following two members override abstract members in Transform */
   val phaseName: String = "constructors"
 
-  protected def newTransformer(unit: CompilationUnit): AstTransformer =
-    new ConstructorTransformer(unit)
+  protected def newTransformer(unit: CompilationUnit): AstTransformer = new ConstructorTransformer(unit)
 
   private val guardedCtorStats: mutable.Map[Symbol, List[Tree]] = perRunCaches.newMap[Symbol, List[Tree]]()
   private val ctorParams: mutable.Map[Symbol, List[Symbol]] = perRunCaches.newMap[Symbol, List[Symbol]]()
@@ -155,8 +154,8 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
    * Finally, the whole affair of eliding is avoided for DelayedInit subclasses,
    * given that for them usually nothing gets elided anyway.
    * That's a consequence from re-locating the post-super-calls statements from their original location
-   * (the primary constructor) into a dedicated synthetic method that an anon-closure may invoke, as required by DelayedInit.
-   *
+   * (the primary constructor) into a dedicated synthetic method that an anon-closure may invoke,
+   * as required by DelayedInit.
    */
   private trait OmittablesHelper {
     def computeOmittableAccessors(clazz: Symbol, defs: List[Tree], auxConstructors: List[Tree], @unused constructor: List[Tree]): Set[Symbol] = {
@@ -337,7 +336,7 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
      * `specializedStats` are replaced by the specialized assignment.
      */
     private def mergeConstructors(genericClazz: Symbol, originalStats: List[Tree], specializedStats: List[Tree]): List[Tree] = {
-      val specBuf = new mutable.ListBuffer[Tree]
+      val specBuf = ListBuffer.empty[Tree]
       specBuf ++= specializedStats
 
       def specializedAssignFor(sym: Symbol): Option[Tree] =
@@ -466,11 +465,15 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
     private val stats = impl.body          // the transformed template body
 
     // find and dissect primary constructor
-    private val (primaryConstr, _primaryConstrParams, primaryConstrBody) = stats collectFirst {
-      case dd@DefDef(_, _, _, vps :: Nil, _, rhs: Block) if dd.symbol.isPrimaryConstructor => (dd, vps map (_.symbol), rhs)
-    } getOrElse {
-      abort("no constructor in template: impl = " + impl)
-    }
+    private val (primaryConstr, _primaryConstrParams, primaryConstrBody) =
+      stats.collectFirst {
+        case dd @ DefDef(_, _, _, vps :: Nil, _, rhs: Block)
+        if dd.symbol.isPrimaryConstructor =>
+          (dd, vps.map(_.symbol), rhs)
+      }
+      .getOrElse {
+        abort("no constructor in template: impl = " + impl)
+      }
 
     def primaryConstrParams  = _primaryConstrParams
     def usesSpecializedField = intoConstructor.usesSpecializedField
@@ -594,7 +597,7 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
       * - `classInitStats`: statements that go into the class initializer
       */
     class Triage {
-      private val defBuf, auxConstructorBuf, constrPrefixBuf, constrStatBuf, classInitStatBuf = new mutable.ListBuffer[Tree]
+      private val defBuf, auxConstructorBuf, constrPrefixBuf, constrStatBuf, classInitStatBuf = ListBuffer.empty[Tree]
 
       triage()
 
@@ -617,8 +620,15 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
           stat match {
             case ValDef(mods, name, _, _) if mods.hasFlag(PRESUPER) => // TODO trait presupers
               // stat is the constructor-local definition of the field value
-              val fields = presupers filter (_.getterName == name)
-              assert(fields.length == 1, s"expected exactly one field by name $name in $presupers of $clazz's early initializers")
+              val fields = presupers.filter { v =>
+                val nm =
+                  if (v.symbol.isPrivateLocal && v.symbol.hasFlag(EXPANDEDNAME))
+                    v.symbol.unexpandedName.dropLocal
+                  else
+                    v.getterName
+                nm == name
+              }
+              assert(fields.length == 1, s"expected exactly one field by name $name in $presupers of $clazz's early initializers but saw $fields")
               val to = fields.head.symbol
 
               if (memoizeValue(to)) constrStatBuf += mkAssign(to, Ident(stat.symbol))
@@ -789,7 +799,7 @@ abstract class Constructors extends Statics with Transform with TypingTransforme
 
       // Eliminate all field/accessor definitions that can be dropped from template
       // We never eliminate delayed hooks or the constructors, so, only filter `defs`.
-      val prunedStats = (defs filterNot omittableStat) ::: delayedHookDefs ::: constructors
+      val prunedStats = defs.filterNot(omittableStat) ::: delayedHookDefs ::: constructors
 
       val statsWithInitChecks =
         if (settings.checkInit.value) {

--- a/test/files/pos/t10561.scala
+++ b/test/files/pos/t10561.scala
@@ -1,0 +1,12 @@
+
+class Parent {
+  private val field: Int = 3
+}
+
+class Child(n: Int) extends {
+  private val field = n
+} with Parent {
+  class Inner {
+    def f = field
+  }
+}

